### PR TITLE
Fix stackoverflow bug in evaluating arithmetic expression which are large lists

### LINKED ISSUE #2278: Crash on bulk update
Hi there,

I'm trying to update the graph using redisgraph-bulk-loader and I get this crash on 2.8.11:

```
=== REDIS BUG REPORT START: Cut & paste starting from here ===
1:M 06 Apr 2022 00:48:31.985 # Redis 6.2.6 crashed by signal: 11, si_code: 2
1:M 06 Apr 2022 00:48:31.985 # Accessing address: 0x7f1a77c8d000
1:M 06 Apr 2022 00:48:31.985 # Crashed running the instruction at: 0x7f1a7c0a7f04

------ STACK TRACE ------
EIP:
/usr/lib/redis/modules/redisgraph.so(+0x1e3f04)[0x7f1a7c0a7f04]

Backtrace:
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12980)[0x7f1a7e48b980]
/usr/lib/redis/modules/redisgraph.so(+0x1e3f04)[0x7f1a7c0a7f04]
/usr/lib/redis/modules/redisgraph.so(AR_EXP_Evaluate+0x98)[0x7f1a7c0a9048]
/usr/lib/redis/modules/redisgraph.so(AR_EXP_ReduceToScalar+0xbf)[0x7f1a7c0a926f]
/usr/lib/redis/modules/redisgraph.so(AR_EXP_FromASTNode+0x22)[0x7f1a7c0a6692]
/usr/lib/redis/modules/redisgraph.so(parse_params+0x16d)[0x7f1a7c0fa4bd]
/usr/lib/redis/modules/redisgraph.so(ExecutionCtx_FromQuery+0x22)[0x7f1a7c0b8332]
/usr/lib/redis/modules/redisgraph.so(_query+0x9d)[0x7f1a7c0b8b4d]
/usr/lib/redis/modules/redisgraph.so(+0x242eb4)[0x7f1a7c106eb4]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x76db)[0x7f1a7e4806db]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7f1a7e1a961f]

------ REGISTERS ------
1:M 06 Apr 2022 00:48:31.995 # 
RAX:00007e72bf58bcfc RBX:00007f1a77879610
RCX:0000000200000000 RDX:00000000000c11c6
RDI:00007e72bf58bcfc RSI:0000000100000008
RBP:00007f1a7848b300 RSP:00007f1a77879610
R8 :00000000000c11c6 R9 :00000000000000f8
R10:0000000000000001 R11:0000000000000003
R12:00000000000413a0 R13:0000000000209cf8
R14:00007e72c6e08090 R15:0000000000000000
RIP:00007f1a7c0a7f04 EFL:0000000000010216
CSGSFS:002b000000000033
1:M 06 Apr 2022 00:48:31.995 # (00007f1a7787961f) -> 0000000200000008
1:M 06 Apr 2022 00:48:31.995 # (00007f1a7787961e) -> 00007e72c6e083fc
1:M 06 Apr 2022 00:48:31.995 # (00007f1a7787961d) -> 0000000200000008
1:M 06 Apr 2022 00:48:31.995 # (00007f1a7787961c) -> 00007e72c6e0839c
1:M 06 Apr 2022 00:48:31.995 # (00007f1a7787961b) -> 0000000200000008
1:M 06 Apr 2022 00:48:31.995 # (00007f1a7787961a) -> 00007e72c6e0833c
1:M 06 Apr 2022 00:48:31.995 # (00007f1a77879619) -> 0000000200000008
1:M 06 Apr 2022 00:48:31.995 # (00007f1a77879618) -> 00007e72c6e082dc
1:M 06 Apr 2022 00:48:31.995 # (00007f1a77879617) -> 0000000200000008
1:M 06 Apr 2022 00:48:31.995 # (00007f1a77879616) -> 00007e72c6e0827c
1:M 06 Apr 2022 00:48:31.995 # (00007f1a77879615) -> 0000000200000008
1:M 06 Apr 2022 00:48:31.995 # (00007f1a77879614) -> 00007e72c6e0821c
1:M 06 Apr 2022 00:48:31.995 # (00007f1a77879613) -> 0000000200000008
1:M 06 Apr 2022 00:48:31.995 # (00007f1a77879612) -> 00007e72c6e081bc
1:M 06 Apr 2022 00:48:31.995 # (00007f1a77879611) -> 0000000200000008
1:M 06 Apr 2022 00:48:31.995 # (00007f1a77879610) -> 00007e72c6e0815c

------ INFO OUTPUT ------
# Server
redis_version:6.2.6
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:74a783698e8a44d9
redis_mode:standalone
os:Linux 5.11.0-1028-oracle x86_64
arch_bits:64
multiplexing_api:epoll
atomicvar_api:c11-builtin
gcc_version:7.5.0
process_id:1
process_supervised:no
run_id:396d8c8375928cf7b9aa75ffd7a972fbdd1c6296
tcp_port:6379
server_time_usec:1649206111917573
uptime_in_seconds:8210
uptime_in_days:0
hz:10
configured_hz:10
lru_clock:5038943
executable:/data/redis-server
config_file:
io_threads_active:0

# Clients
connected_clients:4
cluster_connections:0
maxclients:10000
client_recent_max_input_buffer:18974185
client_recent_max_output_buffer:0
blocked_clients:1
tracking_clients:0
clients_in_timeout_table:0

# Memory
used_memory:325319233744
used_memory_human:302.98G
used_memory_rss:272050577408
used_memory_rss_human:253.37G
used_memory_peak:458860426952
used_memory_peak_human:427.35G
used_memory_peak_perc:70.90%
used_memory_overhead:30861034961
used_memory_startup:844048
used_memory_dataset:294458198783
used_memory_dataset_perc:90.51%
allocator_allocated:325305697408
allocator_active:327506374656
allocator_resident:329744535552
total_system_memory:608295211008
total_system_memory_human:566.52G
used_memory_lua:37888
used_memory_lua_human:37.00K
used_memory_scripts:0
used_memory_scripts_human:0B
number_of_cached_scripts:0
maxmemory:0
maxmemory_human:0B
maxmemory_policy:noeviction
allocator_frag_ratio:1.01
allocator_frag_bytes:2200677248
allocator_rss_ratio:1.01
allocator_rss_bytes:2238160896
rss_overhead_ratio:0.83
rss_overhead_bytes:-57693958144
mem_fragmentation_ratio:0.84
mem_fragmentation_bytes:-53254030832
mem_not_counted_for_evict:0
mem_replication_backlog:0
mem_clients_slaves:0
mem_clients_normal:19056209
mem_aof_buffer:0
mem_allocator:jemalloc-5.1.0
active_defrag_running:0
lazyfree_pending_objects:0
lazyfreed_objects:0

# Persistence
loading:0
current_cow_size:2694750208
current_cow_size_age:22
current_fork_perc:0.53
current_save_keys_processed:2923521
current_save_keys_total:556280002
rdb_changes_since_last_save:1
rdb_bgsave_in_progress:1
rdb_last_save_time:1649197901
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:-1
rdb_current_bgsave_time_sec:25
rdb_last_cow_size:0
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_last_write_status:ok
aof_last_cow_size:0
module_fork_in_progress:0
module_fork_last_cow_size:0

# Stats
total_connections_received:6
total_commands_processed:14
instantaneous_ops_per_sec:0
total_net_input_bytes:20027853
total_net_output_bytes:55814
instantaneous_input_kbps:0.00
instantaneous_output_kbps:0.00
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:0
expired_stale_perc:0.00
expired_time_cap_reached_count:0
expire_cycle_cpu_milliseconds:10
evicted_keys:0
keyspace_hits:556305470
keyspace_misses:0
pubsub_channels:0
pubsub_patterns:0
latest_fork_usec:7540053
total_forks:1
migrate_cached_sockets:0
slave_expires_tracked_keys:0
active_defrag_hits:0
active_defrag_misses:0
active_defrag_key_hits:0
active_defrag_key_misses:0
tracking_total_keys:0
tracking_total_items:0
tracking_total_prefixes:0
unexpected_error_replies:0
total_error_replies:2
dump_payload_sanitizations:0
total_reads_processed:1381
total_writes_processed:15
io_threaded_reads_processed:0
io_threaded_writes_processed:0

# Replication
role:master
connected_slaves:0
master_failover_state:no-failover
master_replid:a68731f9ade00d61fe483db792bc0c7b71edb388
master_replid2:0000000000000000000000000000000000000000
master_repl_offset:0
second_repl_offset:-1
repl_backlog_active:0
repl_backlog_size:1048576
repl_backlog_first_byte_offset:0
repl_backlog_histlen:0

# CPU
used_cpu_sys:366.423267
used_cpu_user:6954.117282
used_cpu_sys_children:0.000938
used_cpu_user_children:0.002828
used_cpu_sys_main_thread:0.563749
used_cpu_user_main_thread:9.959565

# Modules
module:name=graph,ver=20811,api=1,filters=0,usedby=[],using=[],options=[]

# Commandstats
cmdstat_command:calls=2,usec=882,usec_per_call=441.00,rejected_calls=0,failed_calls=0
cmdstat_config:calls=2,usec=32,usec_per_call=16.00,rejected_calls=0,failed_calls=0
cmdstat_graph.QUERY:calls=4,usec=9427,usec_per_call=2356.75,rejected_calls=0,failed_calls=0
cmdstat_ping:calls=1,usec=1,usec_per_call=1.00,rejected_calls=2,failed_calls=0
cmdstat_graph.CONFIG:calls=1,usec=9,usec_per_call=9.00,rejected_calls=0,failed_calls=0
cmdstat_graph.EXPLAIN:calls=2,usec=158,usec_per_call=79.00,rejected_calls=0,failed_calls=0
cmdstat_module:calls=2,usec=8,usec_per_call=4.00,rejected_calls=0,failed_calls=0

# Errorstats
errorstat_LOADING:count=2

# Cluster
cluster_enabled:0

# Keyspace
db0:keys=556280002,expires=0,avg_ttl=0

------ CLIENT LIST OUTPUT ------
id=8 addr=172.17.0.1:54558 laddr=172.17.0.2:6379 fd=12 name= age=7363 idle=1144 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 argv-mem=0 obl=0 oll=0 omem=0 tot-mem=20520 events=r cmd=graph.CONFIG user=default redir=-1
id=10 addr=10.0.0.22:41418 laddr=172.17.0.2:6379 fd=10 name= age=1104 idle=1104 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 argv-mem=0 obl=0 oll=0 omem=0 tot-mem=20512 events=r cmd=graph.QUERY user=default redir=-1
id=11 addr=10.0.0.22:41420 laddr=172.17.0.2:6379 fd=11 name= age=1104 idle=1104 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 argv-mem=0 obl=0 oll=0 omem=0 tot-mem=20512 events=r cmd=graph.QUERY user=default redir=-1
id=21 addr=10.0.0.30:37462 laddr=172.17.0.2:6379 fd=13 name= age=25 idle=10 flags=b db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 argv-mem=18974145 obl=0 oll=0 omem=0 tot-mem=18994665 events=r cmd=graph.QUERY user=default redir=-1

------ MODULES INFO OUTPUT ------
# graph_executing commands
graph_command:GRAPH.QUERY CYPHER rows=[[3194681099,2101678239],[3194689731,2887361942],[3194687104,3181818172],[3194691083,3124004752],[3194686830,3148073094],[3194683418,2043380140],[3194686819,2991450619],[3194680118,1515671919],[3194697409,2059557380],[3194679646,3044883789],[3194682102,3114739377],[3194681799,1276751773],[3194695456,2604481069],[3194688098,2988117198],[3194684428,2037743680],[3194677160,2667058075],[3194681918,2514369987],[3194683706,2032668826],[3194685646,1995895009],[3194680211,3025918553],[3194124418,1965464889],[3194677898,3013392441],[3194691984,2976116439],[3194692128,1602628866],[3194684272,2001598081],[3194689896,2024353402],[3194683871,2054939283],[3194688391,2804912657],[3194685927,1990415494],[3194687734,3179593939],[3194688168,2599980169],[3194688863,2017498566],[3194704800,2728164929],[3193260609,3033457297],[3194688526,2948772047],[3194681461,2052555513],[3194680100,2094691937],[3194684989,2085694387],[3194686651,2071488106],[3194713056,3035891479],[3194682247,2901893832],[3194697723,2857979388],[3194681485,2889188103],[3194687959,2014992247],[3194682655,2050116956],[3194690783,1964848999],[3194690387,3034899291],[3194713097,2125435699],[3194684656,2909487760],[3194688344,2536956629],[3194689018,1802977312],[3194687327,2158497151],[3194682145,2113355106],[3194702207,3035290885],[3194698593,3092051941],[3194693469,2606123402],[3194679582,1996610180],[3194682229,2041205324],[3194660475,3011123403],[3194691268,2534278659],[3194715546,1998088867],[3194698692,2795942599],[3194677771,2584998015],[3194682719,2027636964],[3192234262,2141517543],[3194680845,2026891921],[3194705785,2141125801],[3194688631,1966929797],[3194697082,2889646458],[3194678434,2888575393],[3194717561,2000304796],[3194680847,1990887850],[3194688600,1979624018],[3194697348,2277855494],[3133927089,2756419340],[3194691259,2810517312],[3194701614,2993894543],[3194701539,3000169341],[3194677328,3021339965],[3194687280,2072995355],[3115881678,2092512759],[3194681304,3112931600],[3194697089,2976038876],[3194695234,2179438025],[3194706132,3030292334],[3194677811,3167019627],[3194691926,2134762430],[3194694703,3108218880],[3194684747,3126090857],[3194691166,2099873677],[3194691212,2119194808],[3194688349,1972690852],[3194697484,1967277910],[3194717624,2017254244],[3194686553,2034949102],[3194688408,3040159310],[2928240798,2082057090],[3194684486,3172954106],[3172565675,2957056027],[3194684545,2990118693],[3194684382,1983187697],[3194687817,3165817725],[3194682152,2121210310],[3194682589,2898602234],[3194683838,2045488764],[3189135409,3122052700],[3194693203,2049824182],[3194680546,3132315939],[3194684174,2896178214],[3194716157,2113251333],[3193959556,3146870775],[3194699962,1967294645],[3194687838,2032713966],[3194677619,3193078679],[3194688528,3098838952],[3194690461,2059245554],[3194699540,2560081963],[3082640763,2092673359],[3194693670,2768014158],[3194680958,2092477964],[3194688740,2128768066],[3193431816,2910440302],[3176895987,2530691724],[3194682514,2132525235],[3194692223,2326627678],[3194689727,2046270941],[3194687191,1978891782],[3194682773,2996058680],[3194680319,2313647659],[3194686135,2084743277],[3194688275,3014354229],[3194681699,3030125051],[3194677233,2088726944],[3194699820,2807669085],[3193420102,2077048578],[3194682475,2014247448],[3194687612,2124000937],[3194713626,3189713702],[3194719448,3144491859],[3194685882,1981014184],[3194681846,2036001279],[3194695573,2895178944],[3194677052,2053006289],[3194681133,3000657672],[3194691980,2092136583],[3014373353,2565383127],[3194681637,2789325428],[3194698322,2078270141],[3194684662,1982782877],[3194685747,2951309505],[3194688543,2273765149],[3194677058,2793117440],[3097871001,2554422387],[3194690559,3112376646],[3194704147,2109485845],[3194679287,2035200397],[2133201494,2128252884],[3194715008,1992925035],[3194691047,1968906911],[3194688544,2793790554],[3194682591,3005658178],[3194684375,2909194042],[3194690063,1970677725],[3194688658,2796350009],[3194689025,2783200707],[3194684311,3013372615],[3194684157,2808046661],[3194677200,2524456492],[3194679507,2328237280],[3194689072,1974997758],[3194680959,2000399018],[3194698995,2896548965],[3194680194,1985216588],[3194684155,2465842000],[3194692217,2972512227],[3194691884,3111073034],[3194684691,2765354784],[3194698762,1507402020],[3194688200,2782740965],[2005148041,1614208059],[3194701651,2901698592],[3194692385,2471123203],[3194677102,1970127494],[3194693206,2310818818],[3194688942,1515052776],[3194689945,3126913356],[3194686818,2992291285],[3194685909,2184449996],[3194684255,1997450548],[3194690393,2615895980],[3194681903,1979700870],[3194681094,3033305064],[3194677315,2971797083],[3194689784,3025912000],[3194691272,2036843304],[3194684551,1994910526],[3194682962,2068077411],[3194688010,2567945411],[3194700043,2105484331],[3194689198,3191427491],...(many more rows here)...,[3194686320,3006071760],[2796626703,1964788646],[3194713830,2113273302],[3194693588,3134325013],[3194699217,2159579038],[3194686009,2896073117],[3194692844,2158468653],[3194702149,1885537571],[3194681187,3153724466],[3194689356,3040492409],[3194677653,2117848702],[3194713651,2077230124],[3194683325,3169246879],[3194718693,2103656156],[3194681180,2614968965],[3193959556,3176940474],[3194716635,3100608181],[3194706144,2507676887],[3194698838,3013377779],[3194712838,2983331167],[3194679782,1483042187],[2928240798,2091593512],[3180431925,2793982371],[3194680788,2046332685],[3194715948,2162329773],[3194718253,2769519365],[3194686998,2941241726],[3194679662,2732376772],[3194680542,2548349685],[3194691853,2052771311],[3032843892,2286806038],[3194715509,2057699844],[3194686916,2075951585],[3194706683,2140096025],[3194713153,2157703797],[3194686156,2983641625],[3194683588,3172256280],[3194686757,2139704576],[3194699555,2803908182],[3194690930,2039439639],[3194695411,2012653702],[3194691840,2983644790],[3194679387,2904555264],[3194699653,2009436385],[3194701077,2190424893],[3194712931,2529454491],[3194701546,2478499252],[3194705765,2884669304],[3194698298,2943673605],[3194691750,2087591542],[3193700357,1814307215],[3194698691,2768547285],[3194678521,1994846643],[2057738635,2070735465],[3176957731,2092980791],[3194677038,2076739951],[3194684971,2300804360],[3194687471,2407814661],[3194715157,2118229182],[3194679486,1973044014],[2027532824,2080350957],[3194681206,1971265335],[3194688627,2054385409],[3194680916,2106287765],[3194698610,1980581710],[3194686295,1932477701],[3194693336,2300108605],[3194679573,3122916883],[3168419718,3040183860],[3194715642,2051095471]] UNWIND $rows AS row MATCH (start:Article {id: row[0]}), (end:Article {id: row[1]}) MERGE (start)-[r:CITES]->(end)

------ FAST MEMORY TEST ------
1:M 06 Apr 2022 00:48:32.102 # main thread terminated
1:M 06 Apr 2022 00:48:32.103 # Bio thread for job type #0 terminated
1:M 06 Apr 2022 00:48:32.103 # Bio thread for job type #1 terminated
1:M 06 Apr 2022 00:48:32.103 # Bio thread for job type #2 terminated

Fast memory test PASSED, however your memory can still be broken. Please run a memory test for several hours if possible.

------ DUMPING CODE AROUND EIP ------
Symbol: (null) (base: (nil))
Module: /usr/lib/redis/modules/redisgraph.so (base 0x7f1a7bec4000)
$ xxd -r -p /tmp/dump.hex /tmp/dump.bin
$ objdump --adjust-vma=(nil) -D -b binary -m i386:x86-64 /tmp/dump.bin
------

=== REDIS BUG REPORT END. Make sure to include from START to END. ===
```

Any ideas to work around this?

Thanks.

### PR DESCRIPTION
**Pull request by alonre24**
fixes #2278

### CODE REVIEW

File: src/arithmetic/arithmetic_expression.c [hunk 1]
```diff
@@ -408,6 +411,10 @@ static inline void _AR_EXP_FreeResultsArray(SIValue *results, int count) {
 	for(int i = 0; i < count; i ++) {
 		SIValue_Free(results[i]);
 	}
+    // Large arrays are heap-allocated, so here is where we free it.
```
Comment (AviAvni) [inline src/arithmetic/arithmetic_expression.c:414]: indentation in *.c files in graph is tabs

File: src/arithmetic/arithmetic_expression.c [hunk 2]
```diff
@@ -421,8 +428,14 @@ static AR_EXP_Result _AR_EXP_EvaluateFunctionCall
 	int child_count = node->op.child_count;
 
 	// evaluate each child before evaluating current node
-	SIValue sub_trees[child_count];
-
+    SIValue *sub_trees;
+    // if array size is above the threshold, we allocate it on the heap (otherwise on stack)
+    SIValue sub_trees_on_stack[MAX_ARRAY_SIZE_ON_STACK];
```
Comment (AviAvni) [inline src/arithmetic/arithmetic_expression.c:433]: why always allocate on the stack?
and why always the max num?

**Review by AviAvni — CHANGES_REQUESTED** [review]
see comments

File: src/arithmetic/arithmetic_expression.c (hunk 2 above)
Comment (alonre24) [reply src/arithmetic/arithmetic_expression.c:433]: If I allocate on a stack inside of a block (such as `if` statement) , the variable is no longer valid when we are out of scope.
But after consulting with Roi, I will add a change that will save the unnecessary allocation on the stack.

**Review by swilly22 — APPROVED** [review]

**Review by AviAvni — APPROVED** [review]

**Pull request merged**